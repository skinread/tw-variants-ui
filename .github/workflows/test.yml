# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: CI Tests

on:
  pull_request:
    branches: ["main","develop"]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [20.x]
    steps:
    - name: Checkout
      uses: actions/checkout@v4
    - name: Setup Node ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        cache-dependency-path: "**/package-lock.json"
    - name: Cache or restore dependencies
      uses: actions/cache@v4
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
        restore-keys: ${{ runner.os }}-npm-
    - name: Install dependencies
      run: npm ci

    - name: Check formatting and linting
      run: npm run prettier && npm run lint

    - name: Check package build
      run: npm run build

    - name: Get Playwright version
      run: echo "PLAYWRIGHT_VERSION=$(jq -r '.devDependencies["@playwright/test"]' package.json)" >> $GITHUB_ENV
    - name: Cache or restore Playwright binaries
      uses: actions/cache@v4
      with:
        path: |
          ~/.cache/ms-playwright
        key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
        restore-keys: ${{ runner.os }}-playwright-
    - name: Install Playwright
      if: steps.playwright-cache.outputs.cache-hit != 'true'
      run: npm run pretest
    - name: Storybook test build
      run: npm run storybook:build -- --test
    - name: Storybook test suite
      run: npm run test:ci
