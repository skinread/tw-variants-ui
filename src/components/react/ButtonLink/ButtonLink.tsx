import { buttonVariants } from '@common';
import { Icon, type ButtonProps } from '@components/react';

interface PolymorphicAsProps<E extends React.ElementType> {
  as?: E;
}

// see https://blog.logrocket.com/build-strongly-typed-polymorphic-components-react-typescript/#create-reusable-utility-polymorphic-types
export type PolymorphicProps<E extends React.ElementType> = React.PropsWithChildren<
  React.ComponentPropsWithoutRef<E> & PolymorphicAsProps<E>
>;

type ButtonPropsFiltered = Pick<ButtonProps, 'children' | 'color' | 'fullWidth' | 'className'>;

export interface ButtonLinkProps extends ButtonPropsFiltered {
  /**
   * Destination URL
   */
  href?: string;
  /**
   * Indicate the link goes to an external domain. Important for meeting accessibility best practice by indicating the external nature of the link with an icon
   */
  isExternal?: boolean;
  /**
   * Setting the target is not recommended in production, there are both secuirty and accessibility implications
   */
  target?: React.HTMLAttributeAnchorTarget;
}

const defaultElement = 'a';

type PolymorphicButtonLinkProps<E extends React.ElementType = typeof defaultElement> =
  ButtonLinkProps & PolymorphicProps<E>;

const externalAttrs = {
  rel: 'noopener noreferrer',
  target: '_blank',
};

/**
 * Button styled anchor tag for navigation links. Compose a custom router link (e.g. React Router <Link>) by using the `as` prop. For non navigation buttons, use the standard Button. Default redners as `<a>` tag.
 */
export function ButtonLink<E extends React.ElementType = typeof defaultElement>(
  props: PolymorphicButtonLinkProps<E>
) {
  const {
    as: Component = defaultElement,
    children,
    color,
    fullWidth = false,
    isExternal,
    ...elemAttrs
  } = props;

  const linkAttrs = isExternal ? { ...externalAttrs, ...elemAttrs } : elemAttrs;

  return (
    <Component className={buttonVariants({ color, fullWidth })} {...linkAttrs}>
      {children}
      {!!isExternal && <Icon name="external" width={19} />}
    </Component>
  );
}
