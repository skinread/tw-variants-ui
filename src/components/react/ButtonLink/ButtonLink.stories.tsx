import type { Meta, StoryObj } from '@storybook/react';
import { within, expect, fn } from '@storybook/test';

import { ButtonLink } from './ButtonLink';

const meta = {
  title: 'React/Button Link',
  component: ButtonLink,
  tags: ['autodocs'],
  args: {
    href: '#',
    children: 'Go there',
    color: 'primary',
    fullWidth: false,
    isExternal: undefined,
    onClick: fn(),
  },
} satisfies Meta<typeof ButtonLink>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {};

export const Polymorphic: Story = {
  args: {
    as: 'span',
  },
};

export const External: Story = {
  args: {
    children: <span>Go somehwhere else</span>,
    color: 'secondary',
    fullWidth: true,
    href: 'https://github.com/skinread/',
    isExternal: true,
  },
};

/**
 * Open this story directly to observe automated test
 */
export const InteractonTest: Story = {
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);
    const link = canvas.getByRole('link');
    await expect(link.getAttribute('href')).toBe(args.href);
  },
};
