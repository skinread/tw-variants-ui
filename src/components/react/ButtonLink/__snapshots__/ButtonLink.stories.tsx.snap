// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Button Link External smoke-test 1`] = `
<a class="btn text-base h-auto min-h-fit px-2 py-2 active:bg-primary/30 disabled:opacity-40 disabled:border-primary disabled:text-primary disabled:bg-white btn-block"
   rel="noopener noreferrer"
   target="_blank"
   href="https://github.com/skinread/"
>
  <span>
    Go somehwhere else
  </span>
  <svg xmlns="http://www.w3.org/2000/svg"
       viewbox="0 0 24 24"
       fill="currentColor"
       class="inline-block   icon-external"
       width="19"
       aria-hidden="true"
  >
    <path d="m24,3v7c0,.552-.447,1-1,1s-1-.448-1-1V3.414l-11.793,11.793c-.195.195-.451.293-.707.293s-.512-.098-.707-.293c-.391-.391-.391-1.023,0-1.414L20.586,2h-6.586c-.553,0-1-.448-1-1s.447-1,1-1h7c1.654,0,3,1.346,3,3Zm-1,11c-.553,0-1,.448-1,1v4c0,1.654-1.346,3-3,3H5c-1.654,0-3-1.346-3-3V5c0-1.654,1.346-3,3-3h4c.553,0,1-.448,1-1s-.447-1-1-1h-4C2.243,0,0,2.243,0,5v14c0,2.757,2.243,5,5,5h14c2.757,0,5-2.243,5-5v-4c0-.552-.447-1-1-1Z"
          stroke="none"
    >
    </path>
  </svg>
</a>
`;

exports[`React/Button Link InteractonTest play-test 1`] = `
<a class="btn text-base h-auto min-h-fit px-2 py-2 btn-primary disabled:opacity-40 disabled:bg-primary disabled:text-white"
   href="#"
>
  Go there
</a>
`;

exports[`React/Button Link Polymorphic smoke-test 1`] = `
<span class="btn text-base h-auto min-h-fit px-2 py-2 btn-primary disabled:opacity-40 disabled:bg-primary disabled:text-white"
      href="#"
>
  Go there
</span>
`;

exports[`React/Button Link Primary smoke-test 1`] = `
<a class="btn text-base h-auto min-h-fit px-2 py-2 btn-primary disabled:opacity-40 disabled:bg-primary disabled:text-white"
   href="#"
>
  Go there
</a>
`;
